### ISSUE TITLE
Crash when a labelled-old expression refers to a future label

### ISSUE THREAD

**Issue by vakaras** [viperproject/carbon#366]
This example:
```viper
method test()
{
  var _2: Bool
  _2 := old[l16](true)
  label l16
}
```
gives a stack trace:
```plain
Exception in thread "main" java.util.NoSuchElementException: None.get
        at scala.None$.get(Option.scala:349)
        at scala.None$.get(Option.scala:347)
        at viper.carbon.modules.impls.DefaultExpModule.checkDefinednessImpl(DefaultExpModule.scala:419)
        at viper.carbon.modules.impls.DefaultExpModule.checkDefinedness(DefaultExpModule.scala:322)
        at viper.carbon.modules.impls.DefaultStmtModule.simpleHandleStmt(DefaultStmtModule.scala:88)
        at viper.carbon.modules.impls.DefaultStmtModule.handleStmt(DefaultStmtModule.scala:71)
        at viper.carbon.modules.impls.DefaultStmtModule.$anonfun$translateStmt$5(DefaultStmtModule.scala:270)
        at viper.carbon.modules.impls.DefaultStmtModule.$anonfun$translateStmt$5$adapted(DefaultStmtModule.scala:267)
        at scala.collection.immutable.List.foreach(List.scala:388)
        at scala.collection.generic.TraversableForwarder.foreach(TraversableForwarder.scala:34)
        at scala.collection.generic.TraversableForwarder.foreach$(TraversableForwarder.scala:34)
        at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:43)
        at viper.carbon.modules.impls.DefaultStmtModule.translateStmt(DefaultStmtModule.scala:267)
        at viper.carbon.modules.impls.DefaultStmtModule.$anonfun$translateStmt$3(DefaultStmtModule.scala:246)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:58)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:51)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        at scala.collection.TraversableLike.map(TraversableLike.scala:233)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at viper.carbon.modules.impls.DefaultStmtModule.translateStmt(DefaultStmtModule.scala:246)
        at viper.carbon.modules.impls.DefaultStmtModule.$anonfun$translateStmt$3(DefaultStmtModule.scala:246)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
        at scala.collection.immutable.List.foreach(List.scala:388)
        at scala.collection.TraversableLike.map(TraversableLike.scala:233)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
        at scala.collection.immutable.List.map(List.scala:294)
        at viper.carbon.modules.impls.DefaultStmtModule.translateStmt(DefaultStmtModule.scala:246)
        at viper.carbon.modules.impls.DefaultMainModule.translateMethodDecl(DefaultMainModule.scala:125)
        at viper.carbon.modules.impls.DefaultMainModule.$anonfun$translate$6(DefaultMainModule.scala:80)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:240)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:58)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:51)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:240)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:237)
        at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
        at viper.carbon.modules.impls.DefaultMainModule.translate(DefaultMainModule.scala:80)
        at viper.carbon.CarbonVerifier.verify(CarbonVerifier.scala:148)
        at viper.silver.frontend.DefaultFrontend.verification(Frontend.scala:258)
        at viper.silver.frontend.DefaultFrontend.verification$(Frontend.scala:256)
        at viper.carbon.CarbonFrontend.verification(Carbon.scala:27)
        at viper.silver.frontend.DefaultPhases.$anonfun$Verification$1(Frontend.scala:114)
        at viper.silver.frontend.Frontend.$anonfun$runAllPhases$1(Frontend.scala:60)
        at viper.silver.frontend.Frontend.$anonfun$runAllPhases$1$adapted(Frontend.scala:60)
        at scala.collection.immutable.List.foreach(List.scala:388)
        at viper.silver.frontend.Frontend.runAllPhases(Frontend.scala:60)
        at viper.silver.frontend.Frontend.runAllPhases$(Frontend.scala:59)
        at viper.carbon.CarbonFrontend.runAllPhases(Carbon.scala:27)
        at viper.silver.frontend.SilFrontend.execute(SilFrontend.scala:203)
        at viper.silver.frontend.SilFrontend.execute$(SilFrontend.scala:180)
        at viper.carbon.CarbonFrontend.execute(Carbon.scala:27)
        at viper.carbon.Carbon$.main(Carbon.scala:21)
        at viper.carbon.Carbon.main(Carbon.scala)
```

cc @alexanderjsummers

**Comment by gauravpartha** [comment]
Currently, Carbon deals with labels statically (i.e., must be defined before statically before it is referred to in an old expression). I will adjust the Boogie encoding such that labels can be handled dynamically. In this example, an error should be reported. #128 is a similar issue. The difference is that there the labelled old expression itself is fine, since it is evaluated only if a condition that guarantees the label exists is true.

**Comment by gauravpartha** [comment]
This has been fixed in 2b9ac11.
